### ISSUE TITLE
Docs: setting `ecmaVersion` in `languageOptions` versus `parserOptions`

### ISSUE THREAD

**Issue by akrawitz** [eslint/eslint#20522]
### Docs page(s)

https://eslint.org/docs/latest/use/configure/language-options
https://eslint.org/docs/latest/use/configure/parser

### What documentation issue do you want to solve?

The ESLint docs don't currently make it clear that `ecmaVersion` can be set in either `languageOptions` directly or (at least for the default parser) in the nested `parserOptions` (as documented here: https://github.com/eslint/js/blob/main/packages/espree/README.md, but not on [eslint.org](https://eslint.org/)).

Furthermore, if it has been set in `parserOptions` in one configuration object, then setting it in `languageOptions` in a configuration object further down the cascade does **NOT** override it. This was unexpected and confusing, at least to me.

### What do you think is the correct solution?

I'd like the ESLint docs to highlight this issue of having the `ecmaVersion` setting in two places, and the fact that setting it in `languageOptions` further down the cascade will **NOT** override it if it has been set in `parserOptions` further up the cascade.

This should probably be noted here:
https://eslint.org/docs/latest/use/configure/language-options#specify-parser-options

And here:
https://eslint.org/docs/latest/use/configure/parser#configure-parser-options
The "Tip" at the bottom of this section is particularly confounding since it strongly suggests that `ecmaVersion` always comes from `languageOptions` when in fact it can also be set in `parserOptions` in which case the value passed from `languageOptions` is ignored.

### Participation

- [ ] I am willing to submit a pull request for this change.

### AI acknowledgment

- [x] I did not use AI to generate this issue report.
- [ ] (If the above is not checked) I have reviewed the AI-generated content before submitting.

**Comment by nzakas** [comment]
Thanks for the feedback. I can't find anywhere in the docs where we show an example of setting `ecmaVersion` in `parserOptions`, only in `languageOptions`, which is correct. Given that we never show `parserOptions.ecmaVersion`, I think mentioning it would only encourage more confusion. 

It sounds like you've run into a problem, can you give a concrete example of what you ran into?

**Comment by akrawitz** [comment]
Thanks for your response! Yes, I do have a specific example in mind. I had avoided mentioning it because I was trying to keep things more generally applicable, and because it involves a 3rd-party package which you obviously have no responsibility for, but since you asked:

I was moving from `eslint-config-airbnb-base` to `eslint-config-airbnb-extended`, so that I could move from ESLint v8 and `.eslintrc` to ESLint v10 and flat config.

As a first step, I was using `configs.base.recommended` from `eslint-config-airbnb-extended/legacy`.

Everything was working, except that I kept getting an `espree` parser error on my use of `import.meta.resolve` in a build script. I eventually figured out that this was happening because an earlier `ecmaVersion` was being used. So I tried explicitly setting the `ecmaVersion` to `latest` as a `languageOption` in my local overrides of `eslint-config-airbnb-extended` (in other words, I added it in a configuration object below my inclusion of the Airbnb config). But this had no effect.

Eventually I figured out that the issue was that the Airbnb config sets `ecmaVersion` to `2018` in the `parserOptions` (https://github.com/eslint-config/airbnb-extended/blob/master/packages/eslint-config-airbnb-extended/legacy/configs/base/recommended.ts), which takes precedence over `languageOptions` even if later in the cascade. By setting my override in `parserOptions` I finally fixed the issue.

I submitted this issue in the hope that I could save the next person with a similar issue some time!

**Comment by akrawitz** [comment]
It didn't click for me until I discovered `--print-config` and saw this (with other stuff obviously mixed in):
```js
  "languageOptions": {
    "ecmaVersion": 2026,
    "parserOptions": {
      "ecmaVersion": 2018,
    },
  },
```

**Comment by DMartens** [comment]
The problem of `languageOptions.parserOptions.ecmaVersion` overwriting `languageOptions.ecmaVersion` seems to happen [here](https://github.com/DMartens/eslint/blob/5954ea227745ae484bd0422ae5616044680cec9a/lib/languages/js/index.js#L247-L248).
Should we update the documentation or handle this as a bug-fix?
I think we should handle this as a bug-fix by swapping the order and prefer `ecmaVersion` and `sourceType` from `languageOptions` even if this may be a breaking change.

**Comment by mdjermanovic** [comment]
> The problem of `languageOptions.parserOptions.ecmaVersion` overwriting `languageOptions.ecmaVersion` seems to happen [here](https://github.com/DMartens/eslint/blob/5954ea227745ae484bd0422ae5616044680cec9a/lib/languages/js/index.js#L247-L248). Should we update the documentation or handle this as a bug-fix? I think we should handle this as a bug-fix by swapping the order and prefer `ecmaVersion` and `sourceType` from `languageOptions` even if this may be a breaking change.

It's intentional behavior, per the RFC:

https://github.com/eslint/rfcs/blob/40a07751123998d6437f4b9b1ee311fde7e79089/designs/2019-config-simplification/README.md?plain=1#L94-L95

`languageOptions.ecmaVersion` is used by ESLint internally, e.g., to add ES globals, and is also passed to parsers, but parsers can have their specific values for this option. Setting `languageOptions.parserOptions.ecmaVersion` in config is a way to pass a specific value to a custom parser.

**Comment by akrawitz** [comment]
As an "end user" of ESLint, not having this explained in the docs is more confusing than having it explained, even if the explanation is a bit complicated. I don't know if this "general/specific" pattern only applies to `ecmaVersion` or if it also applies to other settings, but as I now understand it, it works something like this:

> The `ecmaVersion` key has a "general" setting `languageOptions.ecmaVersion` and a more "specific" setting `languageOptions.parserOptions.ecmaVersion`. If the "specific" setting is **never** explicitly assigned, then it defaults to the final value of the "general" setting after all overrides (via extend and cascade) are applied. If the "specific" setting **is** explicitly assigned at **any** point, then the "general" setting has no effect on it, even if the "general" setting is assigned later (via extend or cascade). However, the "specific" setting, once assigned, can be overridden explicitly (via extend or cascade) just as any other setting.

Users need to know which keys this "general/specific" pattern applies to, so that if they try overriding the "general" setting and it does not have the intended effect, they know they might need to override the "specific" setting instead. It seems to me this is an important piece of the overall extend/cascade/override/precedence model.

**Comment by ghost** [comment]
Hi, I'm interested in contributing to ESLint and this issue caught my attention.

From a user's perspective the precedence between
languageOptions.ecmaVersion and parserOptions.ecmaVersion
is indeed surprising.

If the maintainers decide documentation clarification is the
right approach, I'd be happy to help draft the changes.

**Comment by mdjermanovic** [comment]
> This should probably be noted here:
> https://eslint.org/docs/latest/use/configure/language-options#specify-parser-options

This section is specific to the default parser, for which I believe it never makes sense to set `parserOptions.ecmaVersion` or `parserOptions.sourceType` to override `languageOptions.ecmaVersion`/`languageOptions.sourceType`. I think we could add a note saying that these options can have an effect and take precedence over `languageOptions.ecmaVersion` but should never be used.

> And here:
> https://eslint.org/docs/latest/use/configure/parser#configure-parser-options
> The "Tip" at the bottom of this section is particularly confounding since it strongly suggests that `ecmaVersion` always comes from `languageOptions` when in fact it can also be set in `parserOptions` in which case the value passed from `languageOptions` is ignored.

I think we should add more details to the Tip, explaining that it's possible to override `languageOptions.ecmaVersion` and `languageOptions.sourceType` with `parserOptions`.

**Comment by meganprice661-svg** [comment]
Good point! Adding a note about how parserOptions.ecmaVersion can override
languageOptions.ecmaVersion would definitely clear up the confusion,
especially in the "Tip" section.

**Comment by lilianakatrina684-a11y** [comment]
Hi! I’d like to work on this docs clarification and open a PR for it. I’ll keep the change small and focused on the two sections discussed here: clarifying that parserOptions can override languageOptions for ecmaVersion/sourceType, while also making it clear that this shouldn’t be used with the default parser unless you specifically need parser-level behavior. If there’s preferred wording, please let me know. Thanks!

**Comment by vincent067** [comment]
Hi all! 👋

I've been following this issue and wanted to share some thoughts based on my own experience migrating configs.

**The Core Problem**

The confusion stems from the precedence rules not being visually obvious in the cascade. When `parserOptions.ecmaVersion` is set in a shared config (like airbnb's), it silently takes precedence over `languageOptions.ecmaVersion` set later - which feels counterintuitive.

**Documentation Suggestion**

Building on @mdjermanovic's guidance, here's a concrete proposal for the docs:

1. **In `language-options.md`** - Add a ⚠️ warning block in the "Specify Parser Options" section:
   > Warning: Setting `ecmaVersion` or `sourceType` in `parserOptions` will override the values in `languageOptions`. When extending shared configs that set these in `parserOptions`, you may need to override them at the same level.

2. **In `parser.md`** - Update the Tip at the bottom of "Configure Parser Options" to mention:
   > While `languageOptions.ecmaVersion` and `languageOptions.sourceType` are automatically passed to parsers, you can override them per-parser using `parserOptions`. However, be aware that `parserOptions` values take precedence over `languageOptions`.

**Real-World Example**

For anyone debugging this, running `eslint --print-config yourfile.js` is incredibly helpful. You'll see exactly what values are being used:

```json
{
  "languageOptions": {
    "ecmaVersion": 2026,
    "parserOptions": {
      "ecmaVersion": 2018  // This wins! 🏆
    }
  }
}
```

I'd be happy to help review or refine any PR that addresses this. The current state definitely trips people up during ESLint v9 migrations.

Thanks to @akrawitz for the detailed report - saved me debugging time! 🙏

**Comment by nzakas** [comment]
I think we have a decent direction to update the docs here. Marking as accepted.

**Comment by sethamus** [comment]
I’d like to work on this. There’s an open PR (#20880), but it looks like the author can’t sign the CLA.

**Comment by DMartens** [comment]
Feel free to open a PR for this @sethamus.
